### DISCUSSION TITLE
How do I use :iterate in loop?

### DISCUSSION THREAD

**Discussion by uvtc** [janet-lang/janet#588] [General]
What is the purpose of :iterate in a general purpose `loop`, and how do I use it?

If I do:

~~~clojure
(loop [n :range [1 4]
       m :iterate (fn [x] (* x x))]
  (pp (string n "--" m)))
~~~

it forever loops on `n` = 1 with `m` being a different function object each time. What's going on there?

Thanks!

**Comment by pepe** [comment]
Well I am not sure about your usecase, but there is a simple example:
```
(var a 0) (loop [i :iterate (and (> 10 a) (++ a))] (print i))
```
Do not forget it is a macro, so expression is like body for do.

**Comment by uvtc** [comment]
Thanks, @pepe . As for a use case, suppose I want to generate a list of numbers that keep doubling, like 2, 4, 8, 16, ... up until some max-val.

I think I understand that you can use `:iterate` in a `loop` as a binding instead of `:range`.

So, to get that doubling-number list, I tried this:

~~~clojure
(var n 2)

(defn foo
  [x]
  (* x 2))

(loop [n :iterate (foo n)
       :while (< n 1000)]
  (pp n))
~~~

but it's just sticking at n = 4 and endlessly looping. How do I get the variable in the loop to update to the new value of `(foo n)` each time through?

**Comment by sogaiu** [comment]
I think there may be an issue with at least one of the uses of `n` in the `loop` "head" (see the docstring of `loop` for meaning of "head").

Perhaps the following is closer to what you want?

```
(var n 2)

(defn foo [x]
  (* x 2))

(loop [i :iterate (foo n) 
       :while (< n 10)
       :after (set n i)]
  (pp i))
```

**Comment by uvtc** [comment]
Hm. So, the `:iterate` sets `i` (local to the loop) for each iteration of the loop, whereas we update `n` ourselves with `:after` each time through.

**Reply by uvtc** [reply]
I think I was confused because I was expecting `:iterate` to work like Clojure's `iterate` function. But Janet's loop `:iterate` works like, "tell me how to compute the next value for this loop".

I see that, for this use case, I might instead write:

~~~clojure
(var n 2)

(defn foo
  [x]
  (* x 2))

(while (< n 1000)
  (pp n)
  (set n (foo n)))

# or

(for _ 1 10
  (pp n)
  (set n (foo n)))
~~~
